### ISSUE TITLE
Thoughts duplicated when typing fast

### ISSUE THREAD

**Issue by anmolarora1** [cybersemics/em#1071]
The thoughts get duplicated in the state when typing too fast. For some reason, the duplicate thought doesn't show up in the DOM but can be seen in the redux state. 
[Here's the video](https://www.awesomescreenshot.com/video/3274829?key=9dc6c6ad6cd5dfeda1da928875229116) capturing this issue.

**Comment by raineorshine** [comment]
May be related to #973

**Comment by anmolarora1** [comment]
Turns out that the issue is due to the async nature of push and pull middlewares. Everything runs fine as long as the following execution order is maintained - 
`existingThoughtChange` -> `updateThoughts` -> `pushQueue` -> `pullQueue` -> `updateThoughts`
So there are 2 `updateThought` calls for every single update.
First ([existingThoughtChange](https://github.com/cybersemics/em/blob/b8c3236fe54bc0b54a37f1c8b60654eac921af48/src/reducers/existingThoughtChange.ts#L361)) -
```
{
 oldThoughtId: null, 
 newThoughtId: {...newValues}
}
```

Second, [pull action-creator](https://github.com/cybersemics/em/blob/b8c3236fe54bc0b54a37f1c8b60654eac921af48/src/action-creators/pull.ts#L44) 
```
{
 newThoughtId: {...newValues}
}
```

However, when the thoughts are typed in too fast, the push and pull operations from the indexed DB, being async, can't match the speed of updates. As the pull results start lagging, here's what a fetched value looks like - 
```
{
 oldThoughtId: {...oldValues}
}
```
The old thoughts get re-created, hence the duplication.



**Getting to the solution, there can be multiple options, but this is what I'd propose:**

Instead of continuously pulling the thoughts from the indexed DB for all local updates, we should only pull them for the first load/refresh. That'd not just solve the current issue but also help us improve the overall performance of the app by reducing the number of async calls for every thought update.

The only issue that I see with that approach is if the app is running in multiple tabs in the same browser (which is a rare case), but there are ways to identify that and trigger the pull events accordingly. 

I'd like to know your thoughts on this, @raineorshine

**Comment by raineorshine** [comment]
Thanks for investigating and providing the explanation!

> Second, [pull action-creator](https://github.com/cybersemics/em/blob/b8c3236fe54bc0b54a37f1c8b60654eac921af48/src/action-creators/pull.ts#L44)
> 
> ```
> {
>  newThoughtId: {...newValues}
> }
> ```

What is triggering the `pull` after an edit? It just the updated `cursor`? If so, maybe `getVisibleContexts` can ignore the cursor if it is being edited.

> Instead of continuously pulling the thoughts from the indexed DB for all local updates, we should only pull them for the first load/refresh.

Would the subscribe logic effectively do all the pulling needed after initial load?

**Comment by anmolarora1** [comment]
> Thanks for investigating and providing the explanation!
> 
> > Second, [pull action-creator](https://github.com/cybersemics/em/blob/b8c3236fe54bc0b54a37f1c8b60654eac921af48/src/action-creators/pull.ts#L44)
> > ```
> > {
> >  newThoughtId: {...newValues}
> > }
> > ```
> 
> What is triggering the `pull` after an edit? It just the updated `cursor`? If so, maybe `getVisibleContexts` can ignore the cursor if it is being edited.

Based on my understanding so far, it's the `clearPushQueue` action that triggers it. But it's hard to be certain because `pull` is [being triggered](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L104) within a debounced function in the pullQueue [middleware](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L184).


> 
> > Instead of continuously pulling the thoughts from the indexed DB for all local updates, we should only pull them for the first load/refresh.
> 
> Would the subscribe logic effectively do all the pulling needed after initial load?

I'm not sure what you mean by the `subscribe` logic here. Why would we need to subscribe to the local changes?
Can you please help me understand that?

**Comment by raineorshine** [comment]
> > What is triggering the `pull` after an edit? It just the updated `cursor`? If so, maybe `getVisibleContexts` can ignore the cursor if it is being edited.
> 
> Based on my understanding so far, it's the `clearPushQueue` action that triggers it. But it's hard to be certain because `pull` is [being triggered](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L104) within a debounced function in the pullQueue [middleware](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L184).

It might be worth getting clearer on this, but that depends on the answer to the second question. If we only do a `pull` on load, then this finer detail is not needed. If we still have to do a `pull` in some cases after load, it would be good to understand what the additional triggers are.

> > > Instead of continuously pulling the thoughts from the indexed DB for all local updates, we should only pull them for the first load/refresh.
> > 
> > 
> > Would the subscribe logic effectively do all the pulling needed after initial load?
> 
> I'm not sure what you mean by the `subscribe` logic here. Why would we need to subscribe to the local changes?
> Can you please help me understand that?

I'm referring to #932, the task to subscribe to all remote changes. Are you suggesting that we can remove all `pulls` after load because subscribing to the remote will ensure the data is synchronized?

Thanks!

**Comment by anmolarora1** [comment]
> > > What is triggering the `pull` after an edit? It just the updated `cursor`? If so, maybe `getVisibleContexts` can ignore the cursor if it is being edited.
> > 
> > 
> > Based on my understanding so far, it's the `clearPushQueue` action that triggers it. But it's hard to be certain because `pull` is [being triggered](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L104) within a debounced function in the pullQueue [middleware](https://github.com/cybersemics/em/blob/e2c66a7c0b46127f43c2978145b490aed0e0458b/src/redux-middleware/pullQueue.ts#L184).
> 
> It might be worth getting clearer on this, but that depends on the answer to the second question. If we only do a `pull` on load, then this finer detail is not needed. If we still have to do a `pull` in some cases after load, it would be good to understand what the additional triggers are.
> 
> > > > Instead of continuously pulling the thoughts from the indexed DB for all local updates, we should only pull them for the first load/refresh.
> > > 
> > > 
> > > Would the subscribe logic effectively do all the pulling needed after initial load?
> > 
> > 
> > I'm not sure what you mean by the `subscribe` logic here. Why would we need to subscribe to the local changes?
> > Can you please help me understand that?
> 
> I'm referring to #932, the task to subscribe to all remote changes. Are you suggesting that we can remove all `pulls` after load because subscribing to the remote will ensure the data is synchronized?
> 
> Thanks!

Yes, I suggest that we can remove local `pulls` after load. However, it doesn't have anything to do with remote sync (#932). I don't find any reason for subscribing to local changes while the user is editing the thoughts. Let me know if you think otherwise.

**Comment by raineorshine** [comment]
> Yes, I suggest that we can remove local `pulls` after load. However, it doesn't have anything to do with remote sync (#932). I don't find any reason for subscribing to local changes while the user is editing the thoughts. Let me know if you think otherwise.

I think the issue comes up with multi-device use, which has always been a big part of **em** and is something I use every day.

Imagine this scenario:

1. User opens **em** on their phone and loads all thoughts.
2. User opens **em** on their desktop and edits thought A.
3. User goes back to **em** on their phone and edits thought A.

Currently the pull queue will trigger a new `pull` as soon as thought A becomes visible, updating it from the desktop change. If `pull` only occurs once on load, then the local thoughts will not be updated and old data could get submitted to the remote. 

It's related to remote subscribe, since that functionality might be enough to ensure the local data is up-to-date before an edit. I haven't thought through that scenario enough though so I would want to have a strong analysis before removing functionality.

(Ultimately I want CRDT's, but for now we're stuck with my hand-rolled syncing and reconciliation 😊.)

**Comment by anmolarora1** [comment]
> > Yes, I suggest that we can remove local `pulls` after load. However, it doesn't have anything to do with remote sync (#932). I don't find any reason for subscribing to local changes while the user is editing the thoughts. Let me know if you think otherwise.
> 
> I think the issue comes up with multi-device use, which has always been a big part of **em** and is something I use every day.
> 
> Imagine this scenario:
> 
> 1. User opens **em** on their phone and loads all thoughts.
> 2. User opens **em** on their desktop and edits thought A.
> 3. User goes back to **em** on their phone and edits thought A.
> 
> Currently the pull queue will trigger a new `pull` as soon as thought A becomes visible, updating it from the desktop change. If `pull` only occurs once on load, then the local thoughts will not be updated and old data could get submitted to the remote.

I'm not suggesting to pull once on load, but pull **local** once per load/refresh. The remote pull will continue fetching the updates as mentioned in #932

**Comment by raineorshine** [comment]
> I'm not suggesting to pull once on load, but pull **local** (from indexed DB) once per load/refresh. The remote pull will continue fetching the updates as mentioned in #932

I see. So pull IndexedDb once and pull remote on subscribe. 

Currently the app supports editing in multiple tabs. Let's see if we can come up with a solution without losing this useful functionality. Beyond just losing functionality, removing this support could potentially open a new data integrity issue.

What about Dexie's [Observable](https://dexie.org/docs/Observable/Dexie.Observable)? This would allow Dexie to subscribe just like Firebase. That kind of provider agnostic behavior is the intention behind the `DataProvider` protocol. If both IndexedDB and Firebase data can be subscribed to, then I could imagine `pull` only being needed on startup.

---

Another approach is to find a way to identify if a `pull` originated from a local edit. Then they could be filtered out since we know that the data is already in the state. Maybe this is what I was getting at in this TODO in the pull action-creator? I'm not sure though, as I don't remember exactly what `shouldUpdate` is. It does say the consequence is that there are redundant calls to `updateThoughts`.

https://github.com/cybersemics/em/blob/b8c3236fe54bc0b54a37f1c8b60654eac921af48/src/action-creators/pull.ts#L42-L43

Look forward to hearing your thoughts!

**Comment by anmolarora1** [comment]
> > I'm not suggesting to pull once on load, but pull **local** (from indexed DB) once per load/refresh. The remote pull will continue fetching the updates as mentioned in #932
> 
> I see. So pull IndexedDb once and pull remote on subscribe.
> 
> Currently the app supports editing in multiple tabs. Let's see if we can come up with a solution without losing this useful functionality. Beyond just losing functionality, removing this support could potentially open a new data integrity issue.

The multi-tab support can be handled. So there's no loss of functionality. Regarding the potential data integrity issues, let's tackle them if we identify any.


> What about Dexie's [Observable](https://dexie.org/docs/Observable/Dexie.Observable)? This would allow Dexie to subscribe just like Firebase. That kind of provider agnostic behavior is the intention behind the `DataProvider` protocol. If both IndexedDB and Firebase data can be subscribed to, then I could imagine `pull` only being needed on startup.

That'd be possible only if we identify a way to distinguish between locally originated updates(through the same session) vs the rest (other sessions or remote).

> Another approach is to find a way to identify if a `pull` originated from a local edit. Then they could be filtered out since we know that the data is already in the state. Maybe this is what I was getting at in this TODO in the pull action-creator? I'm not sure though, as I don't remember exactly what `shouldUpdate` is. It does say the consequence is that there are redundant calls to `updateThoughts`.
> 

I'd need to do some digging and try to come up with a reasonable solution because even if we disable pulls in the case of a single active session, we'd still need to have them when multiple tabs are open.

**Comment by raineorshine** [comment]
> > What about Dexie's [Observable](https://dexie.org/docs/Observable/Dexie.Observable)? This would allow Dexie to subscribe just like Firebase. That kind of provider agnostic behavior is the intention behind the `DataProvider` protocol. If both IndexedDB and Firebase data can be subscribed to, then I could imagine `pull` only being needed on startup.
> 
> That'd be possible only if we identify a way to distinguish between locally originated updates(through the same session) vs the rest (other sessions or remote).

Yes, I'm guessing this will be necessary to full solve this.

> > Another approach is to find a way to identify if a `pull` originated from a local edit. Then they could be filtered out since we know that the data is already in the state. Maybe this is what I was getting at in this TODO in the pull action-creator? I'm not sure though, as I don't remember exactly what `shouldUpdate` is. It does say the consequence is that there are redundant calls to `updateThoughts`.
> 
> I'd need to do some digging and try to come up with a reasonable solution because even if we disable pulls in the case of a single active session, we'd still need to have them when multiple tabs are open.

Yes, the goal is a general solution that works for any number of tabs open.

Let me know if you'd like to chat or if you'd like to do some digging on your own first!

**Comment by raineorshine** [comment]
> Turns out that the issue is due to the async nature of push and pull middlewares. Everything runs fine as long as the following execution order is maintained -
> `existingThoughtChange` -> `updateThoughts` -> `pushQueue` -> `pullQueue` -> `updateThoughts`
> So there are 2 `updateThought` calls for every single update.

I also just realized that `reconcile` fails to catch the outdated thoughts because there is no local record to compare it to, so it is assumed to be new.

**Comment by anmolarora1** [comment]
> > Turns out that the issue is due to the async nature of push and pull middlewares. Everything runs fine as long as the following execution order is maintained -
> > `existingThoughtChange` -> `updateThoughts` -> `pushQueue` -> `pullQueue` -> `updateThoughts`
> > So there are 2 `updateThought` calls for every single update.
> 
> I also just realized that `reconcile` fails to catch the outdated thoughts because there is no local record to compare it to, so it is assumed to be new. It's a sync problem with the ability to differentiate deletes on one client from adds on another client (and edits in **em** consist of delete + add).

That's correct. I think it's the underlying architecture of having (delete+add) operations instead of just update operations that makes the sync more complex than it should be. I'm not sure if it adds a greater value than the complexities, but I'd like to know your thoughts on that. Do you think we should consider changing it sometime?

**Comment by raineorshine** [comment]
> That's correct. I think it's the underlying architecture of having (delete+add) operations instead of just update operations that makes the sync more complex than it should be. I'm not sure if it adds a greater value than the complexities, but I'd like to know your thoughts on that. Do you think we should consider changing it sometime?

It's one of the biggest mistakes I made in the original architecture. Every thought should have a unique id and we should use that for tracking edits.

**Comment by raineorshine** [comment]
Fixed in #1133 and a4bfc8e63b.
